gateway/src/apicast/policy/rate_limit/README.md

```diff
@@ -0,0 +1,172 @@
+# Rate Limit policy
+
+- [**Description**](#description)
+- [**Types of limits**](#types-of-limits)
+- [**Define limits**](#define-limits)
+- [**Liquid templating**](#liquid-templating)
+- [**Per-gateway vs shared**](#per-gateway-vs-shared)
+- [**Complete config example**](#complete-config-example)
+
+
+## Description
+
+The purpose of this policy is to control the rate of traffic sent to the
+upstream API. There are several kinds of limits that can be defined. Here are
+some examples:
+
+- The client with address 127.0.0.1 can access the specific endpoint
+`/something` up to 100 times per minute.
+- Requests that contain a JSON web token (JWT) whose field `sub` has the value
+`abc` can make up to 10 calls to the API per second.
+
+
+## Types of limits
+
+The types of limits supported by the policy are the ones provided by the
+[lua-resty-limit-traffic](https://github.com/openresty/lua-resty-limit-traffic)
+library:
+
+- Request: based on the "leaky bucket" method.
+- Count: based on a fixed window of time.
+- Connection: based on the concurrent number of connections.
+
+Any limit can be scoped by service or globally.
+
+
+## Define limits
+
+The limits have a key that encodes the entities that we want to use to define
+the limit (an IP, a service, an endpoint, an ID, the value for a specific
+header, etc.). Each limit also has some parameters that vary depending on the
+type:
+
+- Request: `rate`, `burst`. Can make up to `rate` requests per second. It
+allows exceeding that number by `burst` requests per second.
+- Count: `count`, `time_window`. Can make up to `count` requests per
+`time_window` seconds.
+- Connection: `conn`, `burst`, `delay`. `conn` is the max number of concurrent
+connections allowed. It allows exceeding that number by `burst` connections per
+second. `delay` is the number of seconds to delay the requests that exceed the
```

Comment:
`s/requests/connections/`? what happens with connections that are delayed when the other existing connections do not close? (ie. in that case would connections be accepted and permanently over the `conn`/`burst` limit? or would they be delayed again, or closed?)

Reply:
Good question. If I understood correctly, the `lua-resty-limit-traffic` uses a default request latency, and from that, it guesses that a connection will be closed in X seconds in the future. So yes, the prediction can fail and you'd end up with `actual_connections > conn + burst`.

Ref: https://github.com/openresty/lua-resty-limit-traffic/blob/master/README.md

```
-- limit the requests under 200 concurrent requests (normally just
-- incoming connections unless protocols like SPDY is used) with
-- a burst of 100 extra concurrent requests, that is, we delay
-- requests under 300 concurrent connections and above 200
-- connections, and reject any new requests exceeding 300
-- connections.
-- also, we assume a default request time of 0.5 sec, which can be
-- dynamically adjusted by the leaving() call in log_by_lua below.
```